core/src/main/java/io/grpc/internal/ServerImpl.java

```diff
@@ -566,9 +576,72 @@ public void cancelled(Context context) {
 
           context.addListener(new ServerStreamCancellationListener(), directExecutor());
         }
+
+        private <ReqT, RespT> ServerStreamListener processCall(final String fullMethodName,
+                                               final ServerMethodDefinition<ReqT, RespT> methodDef,
+                                               final ServerStream stream,
+                                               final Metadata headers,
+                                               final Context.CancellableContext context,
+                                               final Tag tag) {
+          final ServerCallImpl<ReqT, RespT> call = new ServerCallImpl<>(
+              stream,
+              methodDef.getMethodDescriptor(),
+              headers,
+              context,
+              decompressorRegistry,
+              compressorRegistry,
+              serverCallTracer,
+              tag);
+          return startWrappedCall(fullMethodName, methodDef, call, headers);
+        }
+
+        private <ReqT, RespT> Executor switchExecutorProcessCall(final String fullMethodName,
+            final ServerMethodDefinition<ReqT, RespT> methodDef,
+            final ServerStream stream,
+            final Metadata headers,
+            final Context.CancellableContext context,
+            final Tag tag) {
+          final ServerCallImpl<ReqT, RespT> call = new ServerCallImpl<>(
+                  stream,
+                  methodDef.getMethodDescriptor(),
+                  headers,
+                  context,
+                  decompressorRegistry,
+                  compressorRegistry,
+                  serverCallTracer,
+                  tag);
+          final Executor switchingExecutor = executorSupplier.getExecutor(call, headers);
+          if (switchingExecutor == null) {
+            return null;
+          }
+          final class CallHandled extends ContextRunnable {
+            CallHandled() {
+              super(context);
+            }
+
+            @Override
+            public void runInContext() {
+              ServerStreamListener listener = NOOP_LISTENER;
+              try {
+                listener = startWrappedCall(fullMethodName, methodDef, call, headers);
+              } catch (Throwable t) {
+                stream.close(Status.fromThrowable(t), new Metadata());
+                context.cancel(null);
+                throw t;
+              } finally {
+                jumpListener.setListener(listener);
+                jumpListener.setCallExecutor(defaultExecutor);
+              }
+            }
+          }
+
+          ((SerializingExecutor)defaultExecutor).setExecutor(switchingExecutor);
+          defaultExecutor.execute(new CallHandled());
```

Comment:
In our discussion I thought we decided this approach was broken because this Runnable might be scheduled behind other runnables that are going to use the listener. For example, if the client cancels the call before this line runs.

Although, I realize now a way to resolve that. We can unconditionally enqueue this Runnable at the very beginning.

```java
wrappedExecutor.execute(new LookupMethod()); // creates Call and calls setExecutor() if necessary
wrappedExecutor.execute(new StreamCreated()); // calls jumpListener.setListener()
```

If we want to optimize that, we avoid _LookupMethod_ when unnecessary. Before we were trying to avoid enqueuing the _later_ Runnables, but it seems much easier to avoid enqueuing the earlier runnables (if we care).

The main annoyance is we'll want to pass the created ServerCall to StreamCreated, but we'll have to pass it after the runnable is constructed.

Reply:
> In our discussion I thought we decided this approach was broken because this Runnable might be scheduled behind other runnables that are going to use the listener. For example, if the client cancels the call before this line runs.

It is broken if we execute Runnables that uses the listener immediately. So, we enqueue them all and deliver all at once. But this enqueue needs synchronization that's probably a problem.

> Although, I realize now a way to resolve that. We can unconditionally enqueue this Runnable at the very beginning.
> 
> ```java
> wrappedExecutor.execute(new LookupMethod()); // creates Call and calls setExecutor() if necessary
> wrappedExecutor.execute(new StreamCreated()); // calls jumpListener.setListener()
> ```
> If we want to optimize that, we avoid _LookupMethod_ when unnecessary. Before we were trying to avoid enqueuing the _later_ Runnables, but it seems much easier to avoid enqueuing the earlier runnables (if we care).


Oh this approach sounds nice. i'll try it.
One optimization this approach can't do is to put the second runnable inline of the first one, unless we can cancel the second one before the first one finishes.

> The main annoyance is we'll want to pass the created ServerCall to StreamCreated, but we'll have to pass it after the runnable is constructed.

Reply:
> It is broken if we execute Runnables that uses the listener immediately. So, we enqueue them all and deliver all at once.

Oh, I see it now. I should have looked at that more closely.

> One optimization this approach can't do is to put the second runnable inline of the first one, unless we can cancel the second one before the first one finishes.

SerializingExecutor will do that naturally if the executor doesn't change. Yes, it will be a different Runnable instance, but they are guaranteed to run back-to-back on the same thread (assuming the executor doesn't change).